fhir-server/src/test/java/au/csiro/pathling/fhirpath/function/translate/TranslateFunctionTest.java

```diff
@@ -0,0 +1,446 @@
+        .thenReturn(returnedConceptTranslator);
+
+    // Prepare the inputs to the function.
+    final ParserContext parserContext = new ParserContextBuilder(spark, fhirContext)
+        .idColumn(inputExpression.getIdColumn())
+        .terminologyClientFactory(terminologyClientFactory)
+        .build();
+
+    final StringLiteralPath conceptMapUrlArgument = StringLiteralPath
+        .fromString("'" + CONCEPT_MAP2_URI + "'", inputExpression);
+
+    final BooleanLiteralPath reverseArgument = BooleanLiteralPath
+        .fromString("true", inputExpression);
+
+    final StringLiteralPath equivalenceArgument = StringLiteralPath
+        .fromString("narrower,equivalent", inputExpression);
+
+    final NamedFunctionInput translateInput = new NamedFunctionInput(parserContext, inputExpression,
+        Arrays.asList(conceptMapUrlArgument, reverseArgument, equivalenceArgument));
+    // Invoke the function.
+    final FhirPath result = new TranslateFunction().invoke(translateInput);
+
+    final Dataset<Row> expectedResult = new DatasetBuilder(spark)
+        .withIdColumn()
+        .withEidColumn()
+        .withStructTypeColumns(codingStructType())
+        // TC-1
+        .withRow("encounter-1", makeEid(0, 0), rowFromCoding(TRANSLATED_1))
+        .withRow("encounter-1", makeEid(0, 1), rowFromCoding(TRANSLATED_2))
+        .withRow("encounter-1", makeEid(1, 0), null)
+        .withRow("encounter-1", makeEid(2, 0), rowFromCoding(TRANSLATED_2))
+        // TC-2
+        .withRow("encounter-2", makeEid(0, 0), null)
+        .withRow("encounter-2", makeEid(1, 0), null)
+        // TC-3
+        .withRow("encounter-3", makeEid(0, 0), rowFromCoding(TRANSLATED_1))
+        .withRow("encounter-3", makeEid(0, 1), rowFromCoding(TRANSLATED_2))
+        // TC-4
+        .withRow("encounter-4", makeEid(0, 0), null)
+        // TC-5
+        .withRow("encounter-5", makeEid(0, 0), null)
+        .withRow("encounter-6", null, null)
+        .buildWithStructValue();
+
+    // Check the result.
+    assertThat(result)
+        .hasExpression(
+            "Encounter.type.translate('" + CONCEPT_MAP2_URI + "', true, 'narrower,equivalent')")
+        .isElementPath(CodingPath.class)
+        .hasFhirType(FHIRDefinedType.CODING)
+        .isNotSingular()
+        .selectOrderedResultWithEid()
+        .debugAllRows()
+        .hasRows(expectedResult);
+
+    // Verify mocks
+    final Set<SimpleCoding> expectedSourceCodings = ImmutableSet
+        .of(new SimpleCoding(CODING_1), new SimpleCoding(CODING_2), new SimpleCoding(CODING_3),
+            new SimpleCoding(CODING_4), new SimpleCoding(CODING_5));
+
+    final List<ConceptMapEquivalence> expectedEquivalences = Arrays
+        .asList(ConceptMapEquivalence.NARROWER, ConceptMapEquivalence.EQUIVALENT);
+
+    verify(terminologyService)
+        .translate(eq(expectedSourceCodings), eq(CONCEPT_MAP2_URI), eq(true),
+            eq(expectedEquivalences));
+    verifyNoMoreInteractions(terminologyService);
+  }
```

Comment:
Well it cannot be, because:
```
  public NamedFunctionInput(@Nonnull final ParserContext context,
      @Nonnull final NonLiteralPath input, @Nonnull final List<FhirPath> arguments)
     
```
only accepts inputs that are NonLiteralPaths.

Reply:
Yes, it looks like we check this within `InvocationVisitor#visitFunctionInvocation` - thanks.